ui/test/unit/reports/controllers/reportsController.spec.js

```diff
@@ -108,6 +108,82 @@ describe("ReportsController", function () {
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getDate()).toBe(currentDate.getDate());
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getMonth()).toBe(currentDate.getMonth());
+    });
+
+    it("should return date and month for the this quarter", function () {
+        var currentDate = new Date();
+        rootScope.default.reportsRequiringDateRange = {
+            dateRangeType: new Date(currentDate.getFullYear(), Math.floor(currentDate.getMonth() / 3) * 3, 1)
+        };
+
+        scope.setDefault('dateRangeType', 'reportsRequiringDateRange');
+
+        expect(_.keys(rootScope.default.reportsRequiringDateRange).length).toBe(3);
+        expect((rootScope.default.reportsRequiringDateRange.dateRangeType).getDate()).toBe(1);
+        expect(rootScope.reportsRequiringDateRange[0].startDate.getDate()).toBe(1);
+        expect(rootScope.reportsRequiringDateRange[0].startDate.getMonth()).toBe(currentDate.getMonth());
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getDate()).toBe(currentDate.getDate());
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getMonth()).toBe(currentDate.getMonth());
+    });
+
+    it("should return date and month for the this year", function () {
+        var currentDate = new Date();
+        rootScope.default.reportsRequiringDateRange = {
+            dateRangeType: new Date(currentDate.getFullYear(), 0, 1)
+        };
+
+        scope.setDefault('dateRangeType', 'reportsRequiringDateRange');
+
+        expect(_.keys(rootScope.default.reportsRequiringDateRange).length).toBe(3);
+        expect((rootScope.default.reportsRequiringDateRange.dateRangeType).getDate()).toBe(1);
+        expect(rootScope.reportsRequiringDateRange[0].startDate.getDate()).toBe(1);
+        expect(rootScope.reportsRequiringDateRange[0].startDate.getMonth()).toBe(0);
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getDate()).toBe(currentDate.getDate());
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getMonth()).toBe(currentDate.getMonth());
+    });
+
+    it("should return date and month for last 7 days", function () {
+        var currentDate = new Date();
+        rootScope.default.reportsRequiringDateRange = {
+            dateRangeType: new Date(new Date().setDate(currentDate.getDate() - 6))
+        };
+
+        scope.setDefault('dateRangeType', 'reportsRequiringDateRange');
+
+        expect(_.keys(rootScope.default.reportsRequiringDateRange).length).toBe(3);
+        expect(rootScope.reportsRequiringDateRange[0].startDate.getDate()).toBe(rootScope.default.reportsRequiringDateRange.startDate.getDate());
+        expect(rootScope.reportsRequiringDateRange[0].startDate.getMonth()).toBe(currentDate.getMonth());
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getDate()).toBe(currentDate.getDate());
+        expect(rootScope.reportsRequiringDateRange[0].stopDate.getMonth()).toBe(currentDate.getMonth());
```

Comment:
If the currentDate is `2-Oct-2023 (Mon)`, then for Menu option `Last 7 days`, the dates selected automatically should be: `26-Sep-2023 (Tue)` to `2-Oct-2023 (Mon)`. In this case the startDate month = Sep (and not Oct). But I think your test is checking startDate.getMonth() to be currentDate.getMonth(). That isn't right I think. This test will break on certain days I think. Please reconfirm.

It would have been best if somehow tests could be written independent of actual date (today), and instead could be passed something like: If Current Date is X, then option Last 7 Days should show DateA and DateB, and so on. I am guessing that will require a little refactoring if your code, to take currentDate as a parameter, and then do calculations.

Reply:
Maybe you can otherwise try and use Jasmine clock to mock the current Date. See this ChatGPT discussion I just did regarding this code block (scroll to end for Jasmine clock): https://chat.openai.com/share/42d88843-c43c-4dfc-81bc-6ef80438efbf